decouple workqueue metrics from prometheus

caesarxuchao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
We want to include the workqueue in client-go, but do not want to having to import Prometheus. This PR decouples the workqueue from prometheus.

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #

Partially address https://github.com/kubernetes/kubernetes/issues/33497
User requested for `workqueue` in client-go: https://github.com/kubernetes/client-go/issues/4#issuecomment-249444848

**Special notes for your reviewer**:

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
The implicit registration of Prometheus metrics for workqueue has been removed, and a plug-able interface was added. If you were using workqueue in your own binaries and want these metrics, add the following to your imports in the main package: "k8s.io/pkg/util/workqueue/prometheus".
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33792)

<!-- Reviewable:end -->

Commit:
decouple workqueue metrics from prometheus

krousey commented:
The new public types you're introducing lack documentation, so it's hard to try to infer what they're supposed to be. Like `PartialDefaultMetrics`. Why is it partial, and defaulted? Also you seem to have exposed a factory variable that no one outside of the queue package can effectively use. 

It would be a lot simpler to just have a private interface variable that defaulted to a no-op, and a public registration function. Other than the registration function, and the interface's methods, everything else should be private. e.g.

```go
var metricsProvider MetricsProvider = &noopProvider{}

func SetProvider(p MetricsProvider) { // set once }
```

It also seems like the recording of queue and processing start times are being conflated with the metrics provider here. They should be separated out because one is how you calculate metrics, the other is how you expose and record them.

caesarxuchao commented:
Sorry for the confusion, I'll add the comments.

> Other than the registration function, and the interface's methods, everything else should be private. 

I thought of a better pattern, similar to what you asked for. I'll update the code.

caesarxuchao commented:
@krousey I changed the pattern to similar to what you suggested. I also added comments. Hopefully it's easier to read now.

> It also seems like the recording of queue and processing start times are being conflated with the metrics provider here. They should be separated out because one is how you calculate metrics, the other is how you expose and record them.

The new version should have addressed this comment, if I understood you correctly.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
 	done(item t)
 }
 
+// GaugeMetric that represents a single numerical value that can
```
// GaugeMetric represents ...

caesarxuchao replied:
Done.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
 
+// GaugeMetric that represents a single numerical value that can
+// arbitrarily go up and down.
+// Its methods are a subset of prometheus.Gauge.
```
No need to call out prometheus here. Other implementations could exist.

caesarxuchao replied:
Done.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
-	"github.com/prometheus/client_golang/prometheus"
 )
 
+// You need to import the prometheus package to create and register prometheus
```
Maybe make this more generic and not prometheus specific.

caesarxuchao replied:
Done.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
+	Dec()
+}
+
+type noopGaugeMetric struct{}
```
Your no-ops can be condensed to one type:

```go
type noopMetric struct{}

func (noopMetric) Inc() {}
func (noopMetric) Dec() {}
func (noopMetric) Observe(float64) {}
```

caesarxuchao replied:
Done.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
-		return ret
-	}
+// DepthMetricProvider creates DepthMetric.
+type DepthMetricProvider interface {
```
These should be condensed to the types of metrics they provide, not what the metrics are used for. i.e.

GaugeMetricProvider
ConterMetricProvider
SummaryMetricProvider

caesarxuchao replied:
I condensed them all to a MetricsProvider interface.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -17,63 +17,108 @@ limitations under the License.
-		addTimes:             map[t]time.Time{},
-		processingStartTimes: map[t]time.Time{},
-	}
+type noopDepthMetricProvider struct{}
```
These noop providers can also be condensed to one type.

caesarxuchao replied:
Done.

krousey commented on pkg/util/workqueue/prometheus/prometheus.go:
```diff
@@ -0,0 +1,96 @@
+// prometheus metrics. To use this package, you just have to import it.
+
+func init() {
+	workqueue.DefaultMetricsFactory.SetProviders(
```
Calling a method on a type whose definition is private to another package. Not that it won't work, it just feels wrong. Like if I wanted to search godoc.org for the documentation on `SetProviders`, I wouldn't find it.

caesarxuchao replied:
Thanks for the explanation.

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -151,3 +188,63 @@ func (m *defaultRetryMetrics) retry() {
 	m.retries.Inc()
 }
+
+type metricsFactory struct {
```
I think it might be easiest for me to just show what I meant by not needing a factory variable exposed.

```go
var metricsFactory = struct {
	counterMetricProvider CounterMetricProvider
	gaugeMetricProvider   GaugeMetricProvider
	summaryMetricProvider SummaryMetricProvider
	setProviders          sync.Once
}{
	counterMetricProvider: noopProvider{},
	gaugeMetricProvider:   noopProvider{},
	summaryMetricProvider: noopProvider{},
}

func newQueueMetrics(name string) queueMetrics {
	var ret *defaultQueueMetrics
	if len(name) == 0 {
		return ret
	}
	return &defaultQueueMetrics{
		depth:                metricsFactory.gaugeMetricProvider.NewGuageMetric(name, "depth"),
		adds:                 metricsFactory.counterMetricProvider.NewCounterMetric(name, "adds"),
		latency:              metricsFactory.summaryMetricProvider.NewSummaryMetric(name, "latency"),
		workDuration:         metricsFactory.summaryMetricProvider.NewSummaryMetric(name, "duration"),
		addTimes:             map[t]time.Time{},
		processingStartTimes: map[t]time.Time{},
	}
}

func newRetryMetrics(name string) retryMetrics {
	var ret *defaultRetryMetrics
	if len(name) == 0 {
		return ret
	}
	return &defaultRetryMetrics{
		retries: metricsFactory.counterMetricProvider.NewCounterMetric(name, "retries"),
	}
}

// SetProviders sets the metrics providers of the metricsFactory.
func SetProviders(
	counterMetricProvider CounterMetricProvider,
	gaugeMetricProvider GaugeMetricProvider,
	summaryMetricProvider SummaryMetricProvider,
) {
	metricsFactory.setProviders.Do(func() {
		metricsFactory.counterMetricProvider = counterMetricProvider
		metricsFactory.gaugeMetricProvider = gaugeMetricProvider
		metricsFactory.summaryMetricProvider = summaryMetricProvider
	})
}
```

caesarxuchao replied:
Done

krousey commented on pkg/util/workqueue/prometheus/prometheus.go:
```diff
@@ -0,0 +1,96 @@
+	)
+}
+
+type prometheusDepthMetricProvider struct{}
```
These can also be condensed into one struct with multiple methods.

caesarxuchao replied:
Done

krousey commented on federation/pkg/federation-controller/replicaset/replicasetcontroller.go:
```diff
@@ -43,6 +43,7 @@ import (
 	"k8s.io/kubernetes/pkg/util/flowcontrol"
 	"k8s.io/kubernetes/pkg/util/wait"
 	"k8s.io/kubernetes/pkg/util/workqueue"
+	_ "k8s.io/kubernetes/pkg/util/workqueue/prometheus"
```
I think you try to put this import in package main instead of in these libraries.

caesarxuchao replied:
The new github review UI is confusing. I definitely didn't type "Done" to this comment..

Anyway, do you mean importing it in pkg/controller? I'd rather keep this line next to the workqueue import line for easier tracking.

krousey replied:
I meant importing it in the relevant main packages in cmd/... Like the throttle metrics are done. Perhaps we should consolidate these metrics implementations into one package. Just import that one package and you get prometheus everywhere.

Another PR of course.

caesarxuchao commented:
@k8s-bot unit test this, issue #33838

krousey commented on pkg/util/workqueue/metrics.go:
```diff
@@ -122,32 +113,81 @@ type retryMetrics interface {
+var metricsFactory = struct {
+	metricsProvider MetricsProvider
+	setProviders    sync.Once
+}{}
```
You should initialize `metricsProvider` here to `noopMetricsProvider`.

caesarxuchao replied:
Done.

caesarxuchao commented:
I centralized the imports. I removed the import in the scheduler because it's just using `workqueue.Parallelize()`, so it does not need the metrics. PTAL. Thanks.

caesarxuchao commented:
@k8s-bot e2e test this, issue #33867

caesarxuchao commented:
@krousey can I get the lgtm? Thanks.

caesarxuchao commented:
@lavalamp could you apply the label for this one? I think @krousey is not available.

krousey reviewed (approved):
sorry... was going through email and this was at the bottom of it... until you bumped it that is

caesarxuchao commented:
Thanks @krousey. I just added a release note (copied from yours), let me know if you have comments on that.

caesarxuchao commented:
@k8s-bot node e2e test this, issue #32430
